<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Scope" data-toc-modified-id="Scope-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Scope</a></span></li><li><span><a href="#Import-libraries" data-toc-modified-id="Import-libraries-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Import libraries</a></span></li><li><span><a href="#Importing-CSV-files" data-toc-modified-id="Importing-CSV-files-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Importing CSV files</a></span><ul class="toc-item"><li><span><a href="#2017" data-toc-modified-id="2017-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>2017</a></span></li><li><span><a href="#2018" data-toc-modified-id="2018-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>2018</a></span></li><li><span><a href="#2019" data-toc-modified-id="2019-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>2019</a></span></li><li><span><a href="#2020" data-toc-modified-id="2020-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>2020</a></span></li><li><span><a href="#2021" data-toc-modified-id="2021-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>2021</a></span></li></ul></li><li><span><a href="#Checking-format-of-dfs" data-toc-modified-id="Checking-format-of-dfs-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Checking format of dfs</a></span></li><li><span><a href="#Concatting-2017-and-2018" data-toc-modified-id="Concatting-2017-and-2018-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Concatting 2017 and 2018</a></span><ul class="toc-item"><li><span><a href="#Checking-concatting-format" data-toc-modified-id="Checking-concatting-format-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Checking concatting format</a></span></li><li><span><a href="#Rewording-user_type-contents" data-toc-modified-id="Rewording-user_type-contents-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Rewording <code>user_type</code> contents</a></span></li></ul></li><li><span><a href="#Concatting-2019---2021" data-toc-modified-id="Concatting-2019---2021-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Concatting 2019 - 2021</a></span><ul class="toc-item"><li><span><a href="#Dropping-User-Type" data-toc-modified-id="Dropping-User-Type-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Dropping <code>User Type</code></a></span></li><li><span><a href="#Renaming-columns-to-follow-snake_case" data-toc-modified-id="Renaming-columns-to-follow-snake_case-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Renaming columns to follow snake_case</a></span></li></ul></li><li><span><a href="#Final-DataFrame" data-toc-modified-id="Final-DataFrame-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Final DataFrame</a></span></li><li><span><a href="#Missing-data-EDA" data-toc-modified-id="Missing-data-EDA-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Missing data EDA</a></span></li><li><span><a href="#Dropping-columns-and-rows" data-toc-modified-id="Dropping-columns-and-rows-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Dropping columns and rows</a></span><ul class="toc-item"><li><span><a href="#to_station_id-column" data-toc-modified-id="to_station_id-column-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span><code>to_station_id</code> column</a></span></li><li><span><a href="#from_station_id-column" data-toc-modified-id="from_station_id-column-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span><code>from_station_id</code> column</a></span></li><li><span><a href="#to_station_name-rows" data-toc-modified-id="to_station_name-rows-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span><code>to_station_name</code> rows</a></span></li><li><span><a href="#from_station_name-rows" data-toc-modified-id="from_station_name-rows-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span><code>from_station_name</code> rows</a></span></li><li><span><a href="#user_type-rows" data-toc-modified-id="user_type-rows-9.5"><span class="toc-item-num">9.5&nbsp;&nbsp;</span><code>user_type</code> rows</a></span></li></ul></li><li><span><a href="#trip_duration_seconds-rows" data-toc-modified-id="trip_duration_seconds-rows-10"><span class="toc-item-num">10&nbsp;&nbsp;</span><code>trip_duration_seconds</code> rows</a></span></li><li><span><a href="#Changing-data-types" data-toc-modified-id="Changing-data-types-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Changing data types</a></span></li><li><span><a href="#Exporting-dataframe-as-csv" data-toc-modified-id="Exporting-dataframe-as-csv-12"><span class="toc-item-num">12&nbsp;&nbsp;</span>Exporting dataframe as csv</a></span></li></ul></div>

### Scope
To import individual csv files, clean them, merge them into one main dataframe and then export it as a csv for further analysis.

### Import libraries

In [277]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
from matplotlib.dates import DateFormatter
import matplotlib.dates as mdates
import seaborn as sns
import random 

import warnings
warnings.filterwarnings("ignore")

### Importing CSV files
#### 2017

In [278]:
Y17Q1df = pd.read_csv("2017-Q1.csv")
Y17Q1df.head(1)

trip_id trip_start_time trip_stop_time  trip_duration_seconds  \
0   712382   1/1/2017 0:00  1/1/2017 0:03                    223   

   from_station_id                  from_station_name  to_station_id  \
0             7051  Wellesley St E / Yonge St Green P           7089   

        to_station_name user_type  
0  Church St  / Wood St    Member

In [279]:
Y17Q2df = pd.read_csv("2017-Q2.csv")
Y17Q3df = pd.read_csv("2017-Q3.csv")
Y17Q4df = pd.read_csv("2017-Q4.csv")

In [280]:
# concat
df_2017 = pd.concat([Y17Q1df, Y17Q2df, Y17Q3df, Y17Q4df])

In [281]:
# Check
df_2017

trip_id    trip_start_time     trip_stop_time  trip_duration_seconds  \
0        712382      1/1/2017 0:00      1/1/2017 0:03                    223   
1        712383      1/1/2017 0:00      1/1/2017 0:05                    279   
2        712384      1/1/2017 0:05      1/1/2017 0:29                   1394   
3        712385      1/1/2017 0:07      1/1/2017 0:21                    826   
4        712386      1/1/2017 0:08      1/1/2017 0:12                    279   
...         ...                ...                ...                    ...   
363400  2383642  12/31/17 23:46:27  12/31/17 23:46:53                     26   
363401  2383643  12/31/17 23:47:13   1/01/18 00:11:40                   1467   
363402  2383644  12/31/17 23:47:40  12/31/17 23:57:49                    609   
363403  2383645  12/31/17 23:49:08  12/31/17 23:49:34                     26   
363404  2383646  12/31/17 23:49:41  12/31/17 23:57:41                    480   

        from_station_id                  from_station_name  to_station_id  \
0                7051.0  Wellesley St E / Yonge St Green P         7089.0   
1                7143.0           Kendal Ave / Bernard Ave         7154.0   
2                7113.0       Parliament St / Aberdeen Ave         7199.0   
3                7077.0                 College Park South         7010.0   
4                7079.0              McGill St / Church St         7047.0   
...                 ...                                ...            ...   
363400              NaN           Bloor St / Brunswick Ave            NaN   
363401              NaN           Bloor St / Brunswick Ave            NaN   
363402              NaN            Kendal Ave / Spadina Rd            NaN   
363403              NaN            Phoebe St / Spadina Ave            NaN   
363404              NaN            Phoebe St / Spadina Ave            NaN   

                      to_station_name user_type  
0                Church St  / Wood St    Member  
1             Bathurst Subway Station    Member  
2           College St W / Markham St    Member  
3             King St W / Spadina Ave    Member  
4       University Ave / Gerrard St W    Member  
...                               ...       ...  
363400       Bloor St / Brunswick Ave    Casual  
363401       HTO Park (Queens Quay W)    Casual  
363402       Augusta Ave / Denison Sq    Member  
363403        Phoebe St / Spadina Ave    Member  
363404         Simcoe St / Queen St W    Member  

[1492369 rows x 9 columns]

#### 2018

In [282]:
Y18Q1df = pd.read_csv("2018-Q1.csv")
Y18Q1df.head(1)

trip_id  trip_duration_seconds  from_station_id trip_start_time  \
0  2383648                    393             7018   1/1/2018 0:47   

        from_station_name trip_stop_time  to_station_id  \
0  Bremner Blvd / Rees St  1/1/2018 0:54           7176   

                to_station_name      user_type  
0  Bathurst St / Fort York Blvd  Annual Member

In [283]:
Y18Q2df = pd.read_csv("2018-Q2.csv")
Y18Q3df = pd.read_csv("2018-Q3.csv")
Y18Q4df = pd.read_csv("2018-Q4.csv")

In [284]:
# concat
df_2018 = pd.concat([Y18Q1df, Y18Q2df, Y18Q3df, Y18Q4df])

In [285]:
# Check
df_2018

trip_id  trip_duration_seconds  from_station_id   trip_start_time  \
0       2383648                    393             7018     1/1/2018 0:47   
1       2383649                    625             7184     1/1/2018 0:52   
2       2383650                    233             7235     1/1/2018 0:55   
3       2383651                   1138             7202     1/1/2018 0:57   
4       2383652                    703             7004     1/1/2018 1:00   
...         ...                    ...              ...               ...   
363485  4581273                    379             7088  12/31/2018 23:43   
363486  4581274                    306             7030  12/31/2018 23:45   
363487  4581275                    340             7020  12/31/2018 23:49   
363488  4581276                   1466             7014  12/31/2018 23:52   
363489  4581277                    333             7299  12/31/2018 23:58   

                                 from_station_name    trip_stop_time  \
0                           Bremner Blvd / Rees St     1/1/2018 0:54   
1                       Ossington Ave / College St     1/1/2018 1:03   
2          Bay St / College St (West Side) - SMART     1/1/2018 0:59   
3                 Queen St W / York St (City Hall)     1/1/2018 1:16   
4                          University Ave / Elm St     1/1/2018 1:12   
...                                            ...               ...   
363485                  Danforth Ave / Coxwell Ave  12/31/2018 23:49   
363486                     Bay St / Wellesley St W  12/31/2018 23:50   
363487                     Phoebe St / Spadina Ave  12/31/2018 23:55   
363488  Sherbourne St / Carlton St (Allan Gardens)     1/1/2019 0:17   
363489                     Mill St / Parliament St     1/1/2019 0:04   

        to_station_id                  to_station_name      user_type  
0                7176     Bathurst St / Fort York Blvd  Annual Member  
1                7191       Central Tech  (Harbord St)  Annual Member  
2                7021               Bay St / Albert St  Annual Member  
3                7020          Phoebe St / Spadina Ave  Annual Member  
4                7060      Princess St / Adelaide St E  Annual Member  
...               ...                              ...            ...  
363485           7091                 Donlands Station  Annual Member  
363486           7031          Jarvis St / Isabella St  Annual Member  
363487           7000     Fort York  Blvd / Capreol Ct  Annual Member  
363488           7269  Toronto Eaton Centre (Yonge St)  Annual Member  
363489           7013         Scott St / The Esplanade  Annual Member  

[1922955 rows x 9 columns]

#### 2019

In [286]:
Y19Q1df = pd.read_csv("2019-Q1.csv")
Y19Q1df.head(1)

Trip Id  Trip  Duration  Start Station Id        Start Time  \
0  4581278          1547.0              7021  01/01/2019 00:08   

   Start Station Name  End Station Id          End Time  \
0  Bay St / Albert St            7233  01/01/2019 00:33   

           End Station Name  Bike Id      User Type  
0  King / Cowan Ave - SMART     1296  Annual Member

In [287]:
Y19Q2df = pd.read_csv("2019-Q2.csv")
Y19Q3df = pd.read_csv("2019-Q3.csv")
Y19Q4df = pd.read_csv("2019-Q4.csv")

In [288]:
# concat
df_2019 = pd.concat([Y19Q1df, Y19Q2df, Y19Q3df, Y19Q4df])

In [289]:
# Check
df_2019

Trip Id  Trip  Duration  Start Station Id        Start Time  \
0       4581278          1547.0              7021  01/01/2019 00:08   
1       4581279          1112.0              7160  01/01/2019 00:10   
2       4581280           589.0              7055  01/01/2019 00:15   
3       4581281           259.0              7012  01/01/2019 00:16   
4       4581282           281.0              7041  01/01/2019 00:19   
...         ...             ...               ...               ...   
468411  7334123           523.0              7098  12/31/2019 23:39   
468412  7334124           273.0              7044  12/31/2019 23:45   
468413  7334125          1055.0              7100  12/31/2019 23:51   
468414  7334126           459.0              7470  12/31/2019 23:55   
468415  7334127           361.0              7168  12/31/2019 23:58   

                             Start Station Name  End Station Id  \
0                            Bay St / Albert St          7233.0   
1                      King St W / Tecumseth St          7051.0   
2                        Jarvis St / Carlton St          7013.0   
3       Elizabeth St / Edward St (Bus Terminal)          7235.0   
4                          Edward St / Yonge St          7257.0   
...                                         ...             ...   
468411     Riverdale Park South (Broadview Ave)          7339.0   
468412                 Church St / Alexander St          7273.0   
468413           Dundas St E / Regent Park Blvd          7100.0   
468414              York St / Lake Shore Blvd W          7102.0   
468415                   Queens Quay / Yonge St          7033.0   

                End Time                         End Station Name  Bike Id  \
0       01/01/2019 00:33                 King / Cowan Ave - SMART     1296   
1       01/01/2019 00:29      Wellesley St E / Yonge St (Green P)     2947   
2       01/01/2019 00:25                 Scott St / The Esplanade     2293   
3       01/01/2019 00:20  Bay St / College St (West Side) - SMART      283   
4       01/01/2019 00:24             Dundas St W / St. Patrick St     1799   
...                  ...                                      ...      ...   
468411  12/31/2019 23:48              Carlaw Ave / Strathcona Ave      861   
468412  12/31/2019 23:49              Bay St / Charles St - SMART     3776   
468413  01/01/2020 00:08           Dundas St E / Regent Park Blvd     2382   
468414  01/01/2020 00:03                    Nelson St / Duncan St     2800   
468415  01/01/2020 00:04                            Union Station     4139   

            User Type  
0       Annual Member  
1       Annual Member  
2       Annual Member  
3       Annual Member  
4       Annual Member  
...               ...  
468411  Annual Member  
468412  Annual Member  
468413  Annual Member  
468414  Annual Member  
468415  Annual Member  

[2439517 rows x 10 columns]

#### 2020

In [290]:
Y20M1df = pd.read_csv("2020/2020-01.csv")
Y20M1df.head(1)

Trip Id  Trip  Duration  Start Station Id        Start Time  \
0  7334128             648              7003  01/01/2020 00:08   

         Start Station Name  End Station Id          End Time  \
0  Madison Ave / Bloor St W          7271.0  01/01/2020 00:19   

                  End Station Name  Bike Id      User Type  
0  Yonge St / Alexander St - SMART     3104  Annual Member

In [291]:
# Loading necessary library
from glob import glob

# Setting directory path: "ep_data" is the name of my data file that contains all my necessary files
files = glob("2020/*")
files_rows = [] #empty list to load all my files into

# Loop to open, read and store files into list
for file in files:
    df_2020 = pd.read_csv(file, index_col=None, header=0)
    files_rows.append(df_2020)
    
df_2020 = pd.concat(files_rows, axis=0, ignore_index=True)

In [292]:
# check
df_2020

Trip Id  Trip  Duration Start Station Id        Start Time  \
0         7334128             648             7003  01/01/2020 00:08   
1         7334129             419             7007  01/01/2020 00:10   
2         7334130             566             7113  01/01/2020 00:13   
3         7334131            1274             7333  01/01/2020 00:17   
4         7334132             906             7009  01/01/2020 00:19   
...           ...             ...              ...               ...   
2911303  10644213             330             7010  12/31/2020 23:52   
2911304  10644214             216             7288  12/31/2020 23:54   
2911305  10644215             204             7288  12/31/2020 23:54   
2911306  10644216            1659             7270  12/31/2020 23:56   
2911307  10644217              63             7383  12/31/2020 23:57   

                      Start Station Name End Station Id          End Time  \
0               Madison Ave / Bloor St W         7271.0  01/01/2020 00:19   
1                  College St / Huron St         7163.0  01/01/2020 00:17   
2           Parliament St / Aberdeen Ave         7108.0  01/01/2020 00:22   
3                King St E / Victoria St         7311.0  01/01/2020 00:38   
4                  King St E / Jarvis St         7004.0  01/01/2020 00:34   
...                                  ...            ...               ...   
2911303          King St W / Spadina Ave         7216.0  12/31/2020 23:57   
2911304      Humber Bay Shores Park West         7514.0  12/31/2020 23:58   
2911305      Humber Bay Shores Park West         7514.0  12/31/2020 23:58   
2911306  Church St / Dundas St E - SMART         7270.0  01/01/2021 00:24   
2911307                    12 Harbour St         7383.0  12/31/2020 23:58   

                                  End Station Name Bike Id      User Type  
0                  Yonge St / Alexander St - SMART    3104  Annual Member  
1                               Yonge St / Wood St    2126  Annual Member  
2                           Front St E / Cherry St    4425  Annual Member  
3                      Sherbourne St / Isabella St    4233  Annual Member  
4                          University Ave / Elm St    2341  Casual Member  
...                                            ...     ...            ...  
2911303              Wellington St W / Stafford St  3458.0  Annual Member  
2911304  Humber Bay Shores Park / Marine Parade Dr  4085.0  Annual Member  
2911305  Humber Bay Shores Park / Marine Parade Dr  3580.0  Annual Member  
2911306            Church St / Dundas St E - SMART  5137.0  Annual Member  
2911307                              12 Harbour St  5622.0  Annual Member  

[2911308 rows x 10 columns]

#### 2021

In [293]:
Y21M1df = pd.read_csv("2021/2021-01.csv")
Y21M1df.head(1)

Trip Id  Trip  Duration  Start Station Id        Start Time  \
0  10644218            1315              7021  01/01/2021 00:04   

   Start Station Name  End Station Id          End Time  \
0  Bay St / Albert St          7164.0  01/01/2021 00:26   

                           End Station Name  Bike Id      User Type  
0  Gould St / Yonge St (Ryerson University)   6795.0  Annual Member

In [294]:
# Loading necessary library
from glob import glob

# Setting directory path: "ep_data" is the name of my data file that contains all my necessary files
files = glob("2021/*")
files_rows = [] #empty list to load all my files into

# Loop to open, read and store files into list
for file in files:
    df_2021 = pd.read_csv(file, index_col=None, header=0)
    files_rows.append(df_2021)
    
df_2021 = pd.concat(files_rows, axis=0, ignore_index=True)

In [295]:
# Check
df_2021

Trip Id  Trip  Duration  Start Station Id        Start Time  \
0        10644218            1315              7021  01/01/2021 00:04   
1        10644220             396              7534  01/01/2021 00:07   
2        10644221              86              7162  01/01/2021 00:10   
3        10644222             741              7003  01/01/2021 00:10   
4        10644223            2073              7562  01/01/2021 00:11   
...           ...             ...               ...               ...   
3429328  14635195             359              7121  11/30/2021 23:56   
3429329  14635196            1464              7022  11/30/2021 23:57   
3429330  14635197            1123              7020  11/30/2021 23:57   
3429331  14635198              71              7272  11/30/2021 23:58   
3429332  14635199             451              7026  11/30/2021 23:58   

                           Start Station Name  End Station Id  \
0                          Bay St / Albert St          7164.0   
1                     Walnut Ave / Queen St W          7524.0   
2                    Hayter St / Laplante Ave          7006.0   
3                    Madison Ave / Bloor St W          7272.0   
4        Priscilla  Ave / Dundas St W - SMART          7562.0   
...                                       ...             ...   
3429328               Jarvis St / Dundas St E          7110.0   
3429329                Simcoe St / Queen St W          7001.0   
3429330               Phoebe St / Spadina Ave          7152.0   
3429331       Yonge St / Dundonald St - SMART          7548.0   
3429332                Bay St / St. Joseph St          7129.0   

                 End Time                          End Station Name  Bike Id  \
0        01/01/2021 00:26  Gould St / Yonge St (Ryerson University)   6795.0   
1        01/01/2021 00:13                               Lisgar Park   4176.0   
2        01/01/2021 00:11           Bay St / College St (East Side)   1814.0   
3        01/01/2021 00:22           Yonge St / Dundonald St - SMART    198.0   
4        01/01/2021 00:45      Priscilla  Ave / Dundas St W - SMART   6688.0   
...                   ...                                       ...      ...   
3429328  12/01/2021 00:02                  Queen St E / Berkeley St   4746.0   
3429329  12/01/2021 00:21                 Wellesley Station Green P   1103.0   
3429330  12/01/2021 00:16                Ossington Ave / Bloor St W   6024.0   
3429331  11/30/2021 23:59             St Joseph St / Bay St - SMART    307.0   
3429332  12/01/2021 00:05                  Davenport Rd / Avenue Rd   5146.0   

             User Type  
0        Annual Member  
1        Casual Member  
2        Annual Member  
3        Casual Member  
4        Casual Member  
...                ...  
3429328  Annual Member  
3429329  Casual Member  
3429330  Casual Member  
3429331  Annual Member  
3429332  Casual Member  

[3429333 rows x 10 columns]

### Checking format of dfs

In [296]:
# 2017
df_2017.head(1)

trip_id trip_start_time trip_stop_time  trip_duration_seconds  \
0   712382   1/1/2017 0:00  1/1/2017 0:03                    223   

   from_station_id                  from_station_name  to_station_id  \
0           7051.0  Wellesley St E / Yonge St Green P         7089.0   

        to_station_name user_type  
0  Church St  / Wood St    Member

In [297]:
# 2018
df_2018.head(1)

trip_id  trip_duration_seconds  from_station_id trip_start_time  \
0  2383648                    393             7018   1/1/2018 0:47   

        from_station_name trip_stop_time  to_station_id  \
0  Bremner Blvd / Rees St  1/1/2018 0:54           7176   

                to_station_name      user_type  
0  Bathurst St / Fort York Blvd  Annual Member

In [298]:
# 2019
df_2019.head(1)

Trip Id  Trip  Duration  Start Station Id        Start Time  \
0  4581278          1547.0              7021  01/01/2019 00:08   

   Start Station Name  End Station Id          End Time  \
0  Bay St / Albert St          7233.0  01/01/2019 00:33   

           End Station Name  Bike Id      User Type  
0  King / Cowan Ave - SMART     1296  Annual Member

In [299]:
# 2020
df_2020.head(1)

Trip Id  Trip  Duration Start Station Id        Start Time  \
0  7334128             648             7003  01/01/2020 00:08   

         Start Station Name End Station Id          End Time  \
0  Madison Ave / Bloor St W         7271.0  01/01/2020 00:19   

                  End Station Name Bike Id      User Type  
0  Yonge St / Alexander St - SMART    3104  Annual Member

In [300]:
# 2021
df_2021.head(1)

Trip Id  Trip  Duration  Start Station Id        Start Time  \
0  10644218            1315              7021  01/01/2021 00:04   

   Start Station Name  End Station Id          End Time  \
0  Bay St / Albert St          7164.0  01/01/2021 00:26   

                           End Station Name  Bike Id      User Type  
0  Gould St / Yonge St (Ryerson University)   6795.0  Annual Member

In [301]:
df_2017["user_type"].value_counts()

Member    1164784
Casual     327585
Name: user_type, dtype: int64

In [302]:
df_2018["user_type"].value_counts()

Annual Member    1572980
Casual Member     349975
Name: user_type, dtype: int64

In [303]:
df_2019["User Type"].value_counts()

Annual Member    1859391
Casual Member     580126
Name: User Type, dtype: int64

In [304]:
df_2020["User Type"].value_counts()

Annual Member    1872109
Casual Member    1038950
Name: User Type, dtype: int64

In [305]:
df_2021["User Type"].value_counts()

Annual Member    2032577
Casual Member    1396756
Name: User Type, dtype: int64

`df_2017` and `df_2018` use snake_case where as the remaining dataframes use Title Case. The column `Bike Id` is present starting from 2019. While it can potentially bring in information like wear and tear for example, I will drop it for the sake of simplicity. 

Starting 2018, the column `User Type` has its contents switched from **Member** and **Casual** to **Annual Member** and **Casual Member** respectively. I plan to unify its contents as well.

### Concatting 2017 and 2018

In [306]:
# concat
df_1718 = pd.concat([df_2017, df_2018])

In [307]:
# Check
df_1718

trip_id   trip_start_time    trip_stop_time  trip_duration_seconds  \
0        712382     1/1/2017 0:00     1/1/2017 0:03                    223   
1        712383     1/1/2017 0:00     1/1/2017 0:05                    279   
2        712384     1/1/2017 0:05     1/1/2017 0:29                   1394   
3        712385     1/1/2017 0:07     1/1/2017 0:21                    826   
4        712386     1/1/2017 0:08     1/1/2017 0:12                    279   
...         ...               ...               ...                    ...   
363485  4581273  12/31/2018 23:43  12/31/2018 23:49                    379   
363486  4581274  12/31/2018 23:45  12/31/2018 23:50                    306   
363487  4581275  12/31/2018 23:49  12/31/2018 23:55                    340   
363488  4581276  12/31/2018 23:52     1/1/2019 0:17                   1466   
363489  4581277  12/31/2018 23:58     1/1/2019 0:04                    333   

        from_station_id                           from_station_name  \
0                7051.0           Wellesley St E / Yonge St Green P   
1                7143.0                    Kendal Ave / Bernard Ave   
2                7113.0                Parliament St / Aberdeen Ave   
3                7077.0                          College Park South   
4                7079.0                       McGill St / Church St   
...                 ...                                         ...   
363485           7088.0                  Danforth Ave / Coxwell Ave   
363486           7030.0                     Bay St / Wellesley St W   
363487           7020.0                     Phoebe St / Spadina Ave   
363488           7014.0  Sherbourne St / Carlton St (Allan Gardens)   
363489           7299.0                     Mill St / Parliament St   

        to_station_id                  to_station_name      user_type  
0              7089.0             Church St  / Wood St         Member  
1              7154.0          Bathurst Subway Station         Member  
2              7199.0        College St W / Markham St         Member  
3              7010.0          King St W / Spadina Ave         Member  
4              7047.0    University Ave / Gerrard St W         Member  
...               ...                              ...            ...  
363485         7091.0                 Donlands Station  Annual Member  
363486         7031.0          Jarvis St / Isabella St  Annual Member  
363487         7000.0     Fort York  Blvd / Capreol Ct  Annual Member  
363488         7269.0  Toronto Eaton Centre (Yonge St)  Annual Member  
363489         7013.0         Scott St / The Esplanade  Annual Member  

[3415324 rows x 9 columns]

#### Checking concatting format
Since the 2 dataframes are concat by rows, I can crosscheck the first df_1718 row to df_2017 and the last df_1718 row to df_2018.

In [308]:
df_1718.head(1)

trip_id trip_start_time trip_stop_time  trip_duration_seconds  \
0   712382   1/1/2017 0:00  1/1/2017 0:03                    223   

   from_station_id                  from_station_name  to_station_id  \
0           7051.0  Wellesley St E / Yonge St Green P         7089.0   

        to_station_name user_type  
0  Church St  / Wood St    Member

In [309]:
df_2017.head(1)

trip_id trip_start_time trip_stop_time  trip_duration_seconds  \
0   712382   1/1/2017 0:00  1/1/2017 0:03                    223   

   from_station_id                  from_station_name  to_station_id  \
0           7051.0  Wellesley St E / Yonge St Green P         7089.0   

        to_station_name user_type  
0  Church St  / Wood St    Member

In [310]:
df_1718.tail(1)

trip_id   trip_start_time trip_stop_time  trip_duration_seconds  \
363489  4581277  12/31/2018 23:58  1/1/2019 0:04                    333   

        from_station_id        from_station_name  to_station_id  \
363489           7299.0  Mill St / Parliament St         7013.0   

                 to_station_name      user_type  
363489  Scott St / The Esplanade  Annual Member

In [311]:
df_2018.tail(1)

trip_id  trip_duration_seconds  from_station_id   trip_start_time  \
363489  4581277                    333             7299  12/31/2018 23:58   

              from_station_name trip_stop_time  to_station_id  \
363489  Mill St / Parliament St  1/1/2019 0:04           7013   

                 to_station_name      user_type  
363489  Scott St / The Esplanade  Annual Member

Can confirm that the rows add up.

#### Rewording `user_type` contents

In [312]:
df_1718["user_type"] = np.where(df_1718["user_type"]=="Member","Annual Member", "Casual Member")

In [313]:
df_1718["user_type"].value_counts()

Casual Member    2250540
Annual Member    1164784
Name: user_type, dtype: int64

### Concatting 2019 - 2021

In [314]:
# concat
df_1921 = pd.concat([df_2019, df_2020, df_2021])

In [315]:
# Check
df_1921

Trip Id  Trip  Duration Start Station Id        Start Time  \
0         4581278          1547.0             7021  01/01/2019 00:08   
1         4581279          1112.0             7160  01/01/2019 00:10   
2         4581280           589.0             7055  01/01/2019 00:15   
3         4581281           259.0             7012  01/01/2019 00:16   
4         4581282           281.0             7041  01/01/2019 00:19   
...           ...             ...              ...               ...   
3429328  14635195           359.0             7121  11/30/2021 23:56   
3429329  14635196          1464.0             7022  11/30/2021 23:57   
3429330  14635197          1123.0             7020  11/30/2021 23:57   
3429331  14635198            71.0             7272  11/30/2021 23:58   
3429332  14635199           451.0             7026  11/30/2021 23:58   

                              Start Station Name End Station Id  \
0                             Bay St / Albert St         7233.0   
1                       King St W / Tecumseth St         7051.0   
2                         Jarvis St / Carlton St         7013.0   
3        Elizabeth St / Edward St (Bus Terminal)         7235.0   
4                           Edward St / Yonge St         7257.0   
...                                          ...            ...   
3429328                  Jarvis St / Dundas St E         7110.0   
3429329                   Simcoe St / Queen St W         7001.0   
3429330                  Phoebe St / Spadina Ave         7152.0   
3429331          Yonge St / Dundonald St - SMART         7548.0   
3429332                   Bay St / St. Joseph St         7129.0   

                 End Time                         End Station Name Bike Id  \
0        01/01/2019 00:33                 King / Cowan Ave - SMART    1296   
1        01/01/2019 00:29      Wellesley St E / Yonge St (Green P)    2947   
2        01/01/2019 00:25                 Scott St / The Esplanade    2293   
3        01/01/2019 00:20  Bay St / College St (West Side) - SMART     283   
4        01/01/2019 00:24             Dundas St W / St. Patrick St    1799   
...                   ...                                      ...     ...   
3429328  12/01/2021 00:02                 Queen St E / Berkeley St  4746.0   
3429329  12/01/2021 00:21                Wellesley Station Green P  1103.0   
3429330  12/01/2021 00:16               Ossington Ave / Bloor St W  6024.0   
3429331  11/30/2021 23:59            St Joseph St / Bay St - SMART   307.0   
3429332  12/01/2021 00:05                 Davenport Rd / Avenue Rd  5146.0   

             User Type  
0        Annual Member  
1        Annual Member  
2        Annual Member  
3        Annual Member  
4        Annual Member  
...                ...  
3429328  Annual Member  
3429329  Casual Member  
3429330  Casual Member  
3429331  Annual Member  
3429332  Casual Member  

[8780158 rows x 10 columns]

There are 8780158 rows of data.

#### Dropping `User Type`

In [316]:
# Dropping User Type
df_1921.drop(columns=["Bike Id"], inplace=True)

In [317]:
# Check
df_1921.head()

Trip Id  Trip  Duration Start Station Id        Start Time  \
0  4581278          1547.0             7021  01/01/2019 00:08   
1  4581279          1112.0             7160  01/01/2019 00:10   
2  4581280           589.0             7055  01/01/2019 00:15   
3  4581281           259.0             7012  01/01/2019 00:16   
4  4581282           281.0             7041  01/01/2019 00:19   

                        Start Station Name End Station Id          End Time  \
0                       Bay St / Albert St         7233.0  01/01/2019 00:33   
1                 King St W / Tecumseth St         7051.0  01/01/2019 00:29   
2                   Jarvis St / Carlton St         7013.0  01/01/2019 00:25   
3  Elizabeth St / Edward St (Bus Terminal)         7235.0  01/01/2019 00:20   
4                     Edward St / Yonge St         7257.0  01/01/2019 00:24   

                          End Station Name      User Type  
0                 King / Cowan Ave - SMART  Annual Member  
1      Wellesley St E / Yonge St (Green P)  Annual Member  
2                 Scott St / The Esplanade  Annual Member  
3  Bay St / College St (West Side) - SMART  Annual Member  
4             Dundas St W / St. Patrick St  Annual Member

#### Renaming columns to follow snake_case

In [318]:
df_1921.rename(columns={"Trip Id":"trip_id", "Trip  Duration":"trip_duration_seconds",
                        "Start Station Id":"from_station_id", "Start Time":"trip_start_time", 
                        "Start Station Name":"from_station_name", "End Station Id":"to_station_id", 
                        "End Time":"trip_stop_time", "End Station Name":"to_station_name", 
                        "User Type":"user_type"}, inplace=True)

### Final DataFrame

In [319]:
# concat
BikeShareToronto_df = pd.concat([df_1718, df_1921])

In [320]:
# Check
BikeShareToronto_df

trip_id   trip_start_time    trip_stop_time  trip_duration_seconds  \
0          712382     1/1/2017 0:00     1/1/2017 0:03                  223.0   
1          712383     1/1/2017 0:00     1/1/2017 0:05                  279.0   
2          712384     1/1/2017 0:05     1/1/2017 0:29                 1394.0   
3          712385     1/1/2017 0:07     1/1/2017 0:21                  826.0   
4          712386     1/1/2017 0:08     1/1/2017 0:12                  279.0   
...           ...               ...               ...                    ...   
3429328  14635195  11/30/2021 23:56  12/01/2021 00:02                  359.0   
3429329  14635196  11/30/2021 23:57  12/01/2021 00:21                 1464.0   
3429330  14635197  11/30/2021 23:57  12/01/2021 00:16                 1123.0   
3429331  14635198  11/30/2021 23:58  11/30/2021 23:59                   71.0   
3429332  14635199  11/30/2021 23:58  12/01/2021 00:05                  451.0   

        from_station_id                  from_station_name to_station_id  \
0                7051.0  Wellesley St E / Yonge St Green P        7089.0   
1                7143.0           Kendal Ave / Bernard Ave        7154.0   
2                7113.0       Parliament St / Aberdeen Ave        7199.0   
3                7077.0                 College Park South        7010.0   
4                7079.0              McGill St / Church St        7047.0   
...                 ...                                ...           ...   
3429328            7121            Jarvis St / Dundas St E        7110.0   
3429329            7022             Simcoe St / Queen St W        7001.0   
3429330            7020            Phoebe St / Spadina Ave        7152.0   
3429331            7272    Yonge St / Dundonald St - SMART        7548.0   
3429332            7026             Bay St / St. Joseph St        7129.0   

                       to_station_name      user_type  
0                 Church St  / Wood St  Annual Member  
1              Bathurst Subway Station  Annual Member  
2            College St W / Markham St  Annual Member  
3              King St W / Spadina Ave  Annual Member  
4        University Ave / Gerrard St W  Annual Member  
...                                ...            ...  
3429328       Queen St E / Berkeley St  Annual Member  
3429329      Wellesley Station Green P  Casual Member  
3429330     Ossington Ave / Bloor St W  Casual Member  
3429331  St Joseph St / Bay St - SMART  Annual Member  
3429332       Davenport Rd / Avenue Rd  Casual Member  

[12195482 rows x 9 columns]

### Missing data EDA
The purpose of this section isn't to completely clean data but rather explore and make decisions accordingly. I will most probably delete rows out of convenience mainly because I don't want to fill in information blindly. Example: station ids can potentially change over the years.

In [321]:
BikeShareToronto_df.isna().sum()

trip_id                        0
trip_start_time                0
trip_stop_time                 0
trip_duration_seconds         16
from_station_id          1026893
from_station_name           3313
to_station_id            1030516
to_station_name             7023
user_type                    249
dtype: int64

As of right now, dropping `from_station_id` and `to_station_id` would be very convenient. It's more intuitive to know the name of the stop than its digital counterpart. However, I'll also create a block of code to check if ids of locations were changed over time.

In [322]:
# names of start stations that have missing ids
BikeShareToronto_df[BikeShareToronto_df["from_station_id"].isna()]["from_station_name"].value_counts()

Union Station                                 14675
Simcoe St / Wellington St W                   14176
York St / Queens Quay W                       13451
Dundas St W / Yonge St                        12848
Bay St / Wellesley St W                       11646
                                              ...  
Summerhill Ave / MacLennan Ave - SMART          201
Dundas St W / Roncesvalles Green P - SMART      187
Victoria Park Subway Station - SMART            165
Fringe Next Stage - 7219                         36
Base Station                                      2
Name: from_station_name, Length: 271, dtype: int64

In [323]:
# names of end stations that have missing ids
BikeShareToronto_df[BikeShareToronto_df["to_station_id"].isna()]["to_station_name"].value_counts()

Union Station                                 19239
Simcoe St / Wellington St W                   14741
York St / Queens Quay W                       14570
Dundas St W / Yonge St                        13792
Michael Sweet Ave / St. Patrick St            13141
                                              ...  
Victoria Park Subway Station - SMART            191
Summerhill Ave / MacLennan Ave - SMART          171
Dundas St W / Roncesvalles Green P - SMART      136
Fringe Next Stage - 7219                        125
Base Station                                      2
Name: to_station_name, Length: 271, dtype: int64

Union station seems to have the most missing values.

In [324]:
# number of start IDs union station has (had)
BikeShareToronto_df[BikeShareToronto_df["from_station_name"]=="Union Station"]["from_station_id"].value_counts()

7033.0    70274
7033       1095
Name: from_station_id, dtype: int64

In [325]:
# number of end IDs union station has (had)
BikeShareToronto_df[BikeShareToronto_df["to_station_name"]=="Union Station"]["to_station_id"].value_counts()

7033.0    88960
7033       1275
Name: to_station_id, dtype: int64

Union station in this case seems to have the ID 7033. In this case, I could fill the missing Nans for Union station with 7033. 

**Note:** 7033.0 and 7033 are both and that's due to the concatenation of the data frames while having 2 different data types. 

### Dropping columns and rows

#### `to_station_id` column

In [326]:
BikeShareToronto_df.drop(columns=["to_station_id"], inplace=True)

In [327]:
# Check
BikeShareToronto_df.head(1)

trip_id trip_start_time trip_stop_time  trip_duration_seconds  \
0   712382   1/1/2017 0:00  1/1/2017 0:03                  223.0   

  from_station_id                  from_station_name       to_station_name  \
0          7051.0  Wellesley St E / Yonge St Green P  Church St  / Wood St   

       user_type  
0  Annual Member

#### `from_station_id` column

In [328]:
BikeShareToronto_df.drop(columns=["from_station_id"], inplace=True)

In [329]:
# Check
BikeShareToronto_df.head(1)

trip_id trip_start_time trip_stop_time  trip_duration_seconds  \
0   712382   1/1/2017 0:00  1/1/2017 0:03                  223.0   

                   from_station_name       to_station_name      user_type  
0  Wellesley St E / Yonge St Green P  Church St  / Wood St  Annual Member

#### `to_station_name` rows

In [330]:
BikeShareToronto_df.isna().sum()

trip_id                     0
trip_start_time             0
trip_stop_time              0
trip_duration_seconds      16
from_station_name        3313
to_station_name          7023
user_type                 249
dtype: int64

In [331]:
BikeShareToronto_df.dropna(subset=["to_station_name"],inplace=True)

In [332]:
# Check
BikeShareToronto_df.isna().sum()

trip_id                     0
trip_start_time             0
trip_stop_time              0
trip_duration_seconds      16
from_station_name        3191
to_station_name             0
user_type                 249
dtype: int64

#### `from_station_name` rows

In [333]:
BikeShareToronto_df.dropna(subset=["from_station_name"],inplace=True)

In [334]:
# Check
BikeShareToronto_df.isna().sum()

trip_id                    0
trip_start_time            0
trip_stop_time             0
trip_duration_seconds     16
from_station_name          0
to_station_name            0
user_type                249
dtype: int64

#### `user_type` rows

In [335]:
BikeShareToronto_df.dropna(subset=["user_type"],inplace=True)

In [336]:
BikeShareToronto_df.isna().sum()

trip_id                   0
trip_start_time           0
trip_stop_time            0
trip_duration_seconds    16
from_station_name         0
to_station_name           0
user_type                 0
dtype: int64

### `trip_duration_seconds` rows

In [337]:
BikeShareToronto_df.dropna(subset=["trip_duration_seconds"],inplace=True)

In [338]:
BikeShareToronto_df.isna().sum()

trip_id                  0
trip_start_time          0
trip_stop_time           0
trip_duration_seconds    0
from_station_name        0
to_station_name          0
user_type                0
dtype: int64

In [339]:
BikeShareToronto_df.shape

(12185003, 7)

In [340]:
(8780158/12185003)*100

72.05708525471844

A data retention of ~72.1%. I should definitely get back to this and clean this better (later on) such that I'd retain more information.

### Changing data types

In [341]:
BikeShareToronto_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12185003 entries, 0 to 3429332
Data columns (total 7 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   trip_id                int64  
 1   trip_start_time        object 
 2   trip_stop_time         object 
 3   trip_duration_seconds  float64
 4   from_station_name      object 
 5   to_station_name        object 
 6   user_type              object 
dtypes: float64(1), int64(1), object(5)
memory usage: 743.7+ MB


In [344]:
BikeShareToronto_df["trip_start_time"] = pd.to_datetime(BikeShareToronto_df["trip_start_time"], errors='coerce')

In [345]:
BikeShareToronto_df["trip_stop_time"] = pd.to_datetime(BikeShareToronto_df["trip_stop_time"], errors='coerce')

In [350]:
BikeShareToronto_df["start_date"] = BikeShareToronto_df["trip_start_time"].dt.strftime('%d/%m/%Y')
BikeShareToronto_df["start_time"] = BikeShareToronto_df["trip_start_time"].dt.strftime('%H:%M:%S')
BikeShareToronto_df["end_date"] = BikeShareToronto_df["trip_stop_time"].dt.strftime('%d/%m/%Y')
BikeShareToronto_df["end_time"] = BikeShareToronto_df["trip_stop_time"].dt.strftime('%H:%M:%S')

In [351]:
BikeShareToronto_df

trip_id     trip_start_time      trip_stop_time  \
0          712382 2017-01-01 00:00:00 2017-01-01 00:03:00   
1          712383 2017-01-01 00:00:00 2017-01-01 00:05:00   
2          712384 2017-01-01 00:05:00 2017-01-01 00:29:00   
3          712385 2017-01-01 00:07:00 2017-01-01 00:21:00   
4          712386 2017-01-01 00:08:00 2017-01-01 00:12:00   
...           ...                 ...                 ...   
3429328  14635195 2021-11-30 23:56:00 2021-12-01 00:02:00   
3429329  14635196 2021-11-30 23:57:00 2021-12-01 00:21:00   
3429330  14635197 2021-11-30 23:57:00 2021-12-01 00:16:00   
3429331  14635198 2021-11-30 23:58:00 2021-11-30 23:59:00   
3429332  14635199 2021-11-30 23:58:00 2021-12-01 00:05:00   

         trip_duration_seconds                  from_station_name  \
0                        223.0  Wellesley St E / Yonge St Green P   
1                        279.0           Kendal Ave / Bernard Ave   
2                       1394.0       Parliament St / Aberdeen Ave   
3                        826.0                 College Park South   
4                        279.0              McGill St / Church St   
...                        ...                                ...   
3429328                  359.0            Jarvis St / Dundas St E   
3429329                 1464.0             Simcoe St / Queen St W   
3429330                 1123.0            Phoebe St / Spadina Ave   
3429331                   71.0    Yonge St / Dundonald St - SMART   
3429332                  451.0             Bay St / St. Joseph St   

                       to_station_name      user_type  start_date start_time  \
0                 Church St  / Wood St  Annual Member  01/01/2017   00:00:00   
1              Bathurst Subway Station  Annual Member  01/01/2017   00:00:00   
2            College St W / Markham St  Annual Member  01/01/2017   00:05:00   
3              King St W / Spadina Ave  Annual Member  01/01/2017   00:07:00   
4        University Ave / Gerrard St W  Annual Member  01/01/2017   00:08:00   
...                                ...            ...         ...        ...   
3429328       Queen St E / Berkeley St  Annual Member  30/11/2021   23:56:00   
3429329      Wellesley Station Green P  Casual Member  30/11/2021   23:57:00   
3429330     Ossington Ave / Bloor St W  Casual Member  30/11/2021   23:57:00   
3429331  St Joseph St / Bay St - SMART  Annual Member  30/11/2021   23:58:00   
3429332       Davenport Rd / Avenue Rd  Casual Member  30/11/2021   23:58:00   

           end_date  end_time  
0        01/01/2017  00:03:00  
1        01/01/2017  00:05:00  
2        01/01/2017  00:29:00  
3        01/01/2017  00:21:00  
4        01/01/2017  00:12:00  
...             ...       ...  
3429328  01/12/2021  00:02:00  
3429329  01/12/2021  00:21:00  
3429330  01/12/2021  00:16:00  
3429331  30/11/2021  23:59:00  
3429332  01/12/2021  00:05:00  

[12185003 rows x 11 columns]

### Exporting dataframe as csv

In [352]:
BikeShareToronto_df.to_csv(r"BikeShareToronto_data.csv", index=False, header=True)

In [353]:
BikeShareToronto_df = pd.read_csv("BikeShareToronto_data.csv")
BikeShareToronto_df.head()

trip_id      trip_start_time       trip_stop_time  trip_duration_seconds  \
0   712382  2017-01-01 00:00:00  2017-01-01 00:03:00                  223.0   
1   712383  2017-01-01 00:00:00  2017-01-01 00:05:00                  279.0   
2   712384  2017-01-01 00:05:00  2017-01-01 00:29:00                 1394.0   
3   712385  2017-01-01 00:07:00  2017-01-01 00:21:00                  826.0   
4   712386  2017-01-01 00:08:00  2017-01-01 00:12:00                  279.0   

                   from_station_name                to_station_name  \
0  Wellesley St E / Yonge St Green P           Church St  / Wood St   
1           Kendal Ave / Bernard Ave        Bathurst Subway Station   
2       Parliament St / Aberdeen Ave      College St W / Markham St   
3                 College Park South        King St W / Spadina Ave   
4              McGill St / Church St  University Ave / Gerrard St W   

       user_type  start_date start_time    end_date  end_time  
0  Annual Member  01/01/2017   00:00:00  01/01/2017  00:03:00  
1  Annual Member  01/01/2017   00:00:00  01/01/2017  00:05:00  
2  Annual Member  01/01/2017   00:05:00  01/01/2017  00:29:00  
3  Annual Member  01/01/2017   00:07:00  01/01/2017  00:21:00  
4  Annual Member  01/01/2017   00:08:00  01/01/2017  00:12:00